src/python/pants/jvm/resolve/coursier_setup.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import textwrap
+from dataclasses import dataclass
+
+from pants.core.util_rules.external_tool import (
+    DownloadedExternalTool,
+    ExternalToolRequest,
+    TemplatedExternalTool,
+)
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.platform import Platform
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+
+logger = logging.getLogger(__name__)
+
+
+COURSIER_POST_PROCESSING_SCRIPT = textwrap.dedent(
+    """\
+    import json
+    import sys
+    from pathlib import PurePath
+    from shutil import copyfile
+
+    report = json.load(open(sys.argv[1]))
+
+    classpath = set()
+    for dep in report['dependencies']:
+        file_path = PurePath(dep['file'])
+        classpath_dest = f"classpath/{file_path.name}"
+        if classpath_dest in classpath:
+            raise Exception(f"Found duplicate jar name {file_path.name}, which isn't currently supported")
+        classpath.add(classpath_dest)
+        copyfile(file_path, classpath_dest)
+    """
+)
+
+COURSIER_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+    set -eux
+
+    coursier_exe="$1"
+    shift
+    json_output_file="$1"
+    shift
+
+    "$coursier_exe" fetch --json-output-file="$json_output_file" "$@"
+
+    /bin/mkdir -p classpath
+    /usr/bin/python3 coursier_post_processing_script.py "$json_output_file"
+    """
+)
+
+JAVAC_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+    set -eux
+
+    coursier_exe="$1"
+    shift
+
+    javac_path="$(${coursier_exe} java-home --jvm openjdk:1.9.0-4)/bin/javac"
+
+    exec "${javac_path}" "$@"
+    """
+)
+
+
+class CoursierBinary(TemplatedExternalTool):
+    options_scope = "coursier"
+    help = "A dependency resolver for the Maven ecosystem."
+
+    default_version = "v2.0.13"
+    default_known_versions = [
+        "v2.0.13|linux|1ae089789cc4b0a4d296d6852b760d7f8bf72805267a6b7571e99b681d5e13b4|59652208",
+        "v2.0.13|darwin|d74b8fe4ffc2f4e9011d7151722fc8b5ffca8a72b3bc4188c61df3326228c4ef|57625024",
+    ]
+    default_url_template = (
+        "https://github.com/coursier/coursier/releases/download/{version}/cs-{platform}"
+    )
+    default_url_platform_mapping = {"darwin": "x86_64-apple-darwin", "linux": "x86_64-pc-linux"}
+
+
+@dataclass(frozen=True)
+class Coursier:
+    """The Coursier tool and various utilities, materialzed to a `Digest` and ready to use."""
+
+    coursier: DownloadedExternalTool
+    digest: Digest
+    wrapper_script: str = "coursier_wrapper_script.sh"
+    post_processing_script: str = "coursier_post_processing_script.py"
+    javac: str = "coursier_javac.sh"
```

Comment:
This works, but better is to use `ClassVar[str]` for each of these, from `typing` module. https://docs.python.org/3/library/dataclasses.html#class-variables